source/neuropod/tests/test_utils.hh

```diff
@@ -209,3 +209,62 @@ void test_strings_model_ope(const std::string &neuropod_path)
     neuropod::Neuropod neuropod(neuropod_path, opts);
     test_strings_model(neuropod);
 }
+
+void test_torchscript_dict_with_union_value_type_model(neuropod::Neuropod &neuropod)
+{
+#if CAFFE2_NIGHTLY_VERSION >= 20220127
```

Comment:
1. This file is meant for generic test utils - not ones that are specific to a particular backend. These additions should be either renamed to be more generic (if we can use them in tests for other backends as well) or moved into the torch backend folder.

2. `CAFFE2_NIGHTLY_VERSION` is defined in `torch_tensor.hh`: https://github.com/uber/neuropod/blob/1097df8fc683141db08f8d66c67cae45d0a83a33/source/neuropod/backends/torchscript/torch_tensor.hh#L32-L80 Since we're not including `torch_tensor.hh` here or in `test_torchscript_backend.cc`, `CAFFE2_NIGHTLY_VERSION` is undefined. This means none of your gated test code is included in the build.

A few solutions:

* Rename the functions to be backend agnostic and move the `#if CAFFE2_...` into torch specific files (which include `torch_tensor.hh`)
* Move these functions into the torchscript backend test directory and include `torch_tensor.hh`

To make problems like this easier to catch, we should also build with `-Wundef`. I had a series of PRs converting the Neuropod build to use a very strict set of warnings flags (see [here](https://github.com/uber/neuropod/blob/1097df8fc683141db08f8d66c67cae45d0a83a33/source/bazel/cc.bzl#L4-L26)), but those settings are only used by part of the build at the moment.